### PR TITLE
Enhance GuestMemory to support memory hotplug 

### PR DESCRIPTION
**Pull request by bonzini**
Since @jiangliu closed #64, and this code is now documented and I think in decent shape, let's create a draft pull request to give it more exposure.

The main difference with #64 is that I am touching the core traits to allow:

* a generic implementation a memory map that wraps another `GuestMemory` M and adds RCU-style mutability (`GuestMemoryAtomic<M>`);

* no panics and no fighting the borrow checker for `find_region`, which makes no sense for a mutable memory map (see the `GuestMemoryMut`/`GuestMemory` split in the first patch)

* an abstraction of "snapshotting" a memory map that is zero-cost when the memory map is immutable (`GuestMemorySnapshottable<M>`). That is, whenever client code (e.g. vm-virtio) adds support for hotplug, it will not impose any penalty on VMMs whose memory map is immutable and passed around as e.g. `&'a GuestMemoryMmap`.

Even though it's very different from #64, it is definitely inspired by it and the concepts are entirely an evolution of what @jiangliu had proposed.

This is definitely pushing my Rust knowledge and confidence to the limits, so I'm looking forward to the reviews!

### CODE REVIEW

File: src/mmap.rs [hunk 1]
```diff
@@ -444,6 +444,62 @@ impl GuestMemoryMmap {
...
+        // It shouldn't be a huge Vec, so prefer simple implementation.
+        for reg in self.regions.iter() {
+            if region.start_addr() >= reg.start_addr() && region.start_addr() <= reg.last_addr() {
+                return Err(Error::MemoryRegionOverlap);
+            } else if reg.start_addr() >= region.start_addr()
+                && reg.start_addr() <= region.last_addr()
+            {
+                return Err(Error::MemoryRegionOverlap);
+            }
+        }
```
Comment (serban300) [inline src/mmap.rs]: Since some similar logic is also used in `GuestMemoryMmap::from_arc_regions()` I think it would be worth defining a method that checks if 2 regions overlap inside the `GuestMemoryRegion` trait.

**Comment by rbradford** [comment]
Hi @bonzini this looks pretty good and is a more generalised version of the approach I took to implement hotplug in Cloud Hypervisor. I think the next step is probably to see if I can port our code over to use it as that will highlight any issues with this PR or with other bits of the tree (i'm kinda worried about linux-loader and some of our devices.)

**Comment by bonzini** [comment]
Yes, the generalization ensures zero cost for immutable memory maps.

Should I separate the first patch into its own pull request already? Be sure to take a look  at the docs because they explain an associated type trick that can come in handy when porting vm-virtio to this extension.

File: src/guest_memory.rs [hunk 2]
```diff
@@ -302,6 +306,96 @@ pub trait GuestMemoryMut {
...
+/// impl<'a> GuestMemoryBackend<'a> for ImmutableMmap {
+///     type Snapshot = GuestMemoryMmap;
+///     type Memory = &'a GuestMemoryMmap;
+/// }
+/// let mut for_immutable_mmap: VirtioDevice<ImmutableMmap> =
+///     VirtioDevice::new();
+/// let mmap = get_mmap();
+/// for_immutable_mmap.activate(&mmap);
+/// # }
+/// ```
```
Comment (jiangliu) [inline src/guest_memory.rs]: In my original design, I used `M: GuestMemory` for both GuestMemoryBackend::Snapshot and GuestMemoryBackend::Memory.
The consideration is that:
1) for GuestMemory clients not caring about performance and only using stateless interfaces such as Bytes, it doesn't need to care about memory hotplug and snapshot().
2) for GuestMemory clients caring about performance or using stateful interfaces such as find_region/num_region/last_addr etc, it need to use the snapshot() method to get a stable GuestMemory snapshot.

With this proposed design, we need to teach all GuestMemory consumers about memory hotplug and snapshot(). But I'm on this side to clearly expose Memory/Snapshot concepts to vm-memory consumers.

And the changes to vm-virtio and virtio device drivers should be acceptable.

**Comment by jiangliu** [comment]
> Yes, the generalization ensures zero cost for immutable memory maps.
> 
> Should I separate the first patch into its own pull request already? Be sure to take a look at the docs because they explain an associated type trick that can come in handy when porting vm-virtio to this extension.

Seems we are almost on the same page, any comments from @alexandruag ?
As both Cloud hypervisor and we are both working on memory hotplug, it would be great to merge all together:)

**Comment by jiangliu** [comment]
We should be great to merge this part before publishing to crates.io

File: src/guest_memory.rs [hunk 3]
```diff
@@ -309,6 +304,134 @@ pub trait GuestMemory {
...
+/// to repetitive types such as `VirtioDevice<GuestMemoryMmap, &GuestMemoryMmap>`.
+///
+/// ```
+/// # #[cfg(feature = "backend-mmap")]
+/// # use vm_memory::GuestMemoryMmap;
+/// # #[cfg(feature = "backend-atomic")]
+/// # use vm_memory::GuestMemoryAtomic;
+/// # use vm_memory::{GuestAddress, GuestMemory, GuestMemorySnapshottable};
+///
+/// # // Perhaps this could be added to vm-memory itself?
```
Comment (jiangliu) [inline src/guest_memory.rs]: Definitely it should be included in vm-memory, because it's may be the main interfaces used by consumers such as vm-virtio.

File: src/mmap.rs (hunk 1 above)
Comment (bonzini) [reply src/mmap.rs]: Good idea.

File: src/guest_memory.rs (hunk 2 above)
Comment (bonzini) [reply src/guest_memory.rs]: > for GuestMemory clients not caring about performance and only using stateless interfaces such as Bytes, it doesn't need to care about memory hotplug and snapshot().

The main issue is that you could have a client caring about performance and potentially hotplug (`vm-virtio`), but it should not add any cost if the `GuestMemory` user does not need hotplug (like firecracker).

> we need to teach all GuestMemory consumers about memory hotplug and snapshot(). 

Not all of them. You can keep using `GuestMemory` just like now if you don't support hotplug. The need for changes only arises if you want to use `GuestMemoryMut` to support hotplug. Then you need to introduce `snapshot()`.

Comment (jiangliu) [reply src/guest_memory.rs]: > > for GuestMemory clients not caring about performance and only using stateless interfaces such as Bytes, it doesn't need to care about memory hotplug and snapshot().
> 
> The main issue is that you could have a client caring about performance and potentially hotplug (`vm-virtio`), but it should not add any cost if the `GuestMemory` user does not need hotplug (like firecracker).
There are two cases here,
1) vmm supports memory hotplug or not. If the vmm doesn't support hotplug, it's easy to handle, current GuestMemory trait just works without code changes.
2) components (such as device drivers) support memory hotplug or not when the vmm supports memory hotplug.
What I'm thinking is about case 2.

> 
> > we need to teach all GuestMemory consumers about memory hotplug and snapshot().
> 
> Not all of them. You can keep using `GuestMemory` just like now if you don't support hotplug. The need for changes only arises if you want to use `GuestMemoryMut` to support hotplug. Then you need to introduce `snapshot()`.

The issue arises when sharing code among VMMs. For VMM specific implementations, we could just choose one of hotplug or non-hotplug interfaces. To share code, such as vm-device/vm-virtio, we need to teach them about hotplug/snapshot().

**Comment by alexandruag** [comment]
> Seems we are almost on the same page, any comments from @alexandruag ?
> As both Cloud hypervisor and we are both working on memory hotplug, it would be great to merge all together:)

Sorry, I'm still trying to warp my head around a couple of things. Will post any comments by Monday EOD.

**Comment by sboeuf** [comment]
@bonzini could you please rebase on top of master? Mainly you're missing the latest changes from https://github.com/rust-vmm/vm-memory/pull/70

File: src/guest_memory.rs [hunk 4]
```diff
@@ -302,6 +306,96 @@ pub trait GuestMemoryMut {
...
+/// for_immutable_mmap.activate(&mmap);
+/// # }
+/// ```
+
+pub trait GuestMemorySnapshottable<M: GuestMemory> {
+    /// The type of the object returned by `snapshot()`.
+    type S: Deref<Target = M>;
+
+    /// Return a consistent snapshot of the memory map.
+    fn snapshot(&self) -> Self::S;
```
Comment (alexandruag) [inline src/guest_memory.rs]: The biggest question in my mind right now is whether this method needs to be a first class citizen of the `GuestMemory` interface. I can think of two potential reasons:
- There are use cases where the region set needs to remain immutable over the course of multiple `GuestMemory` method invocations. Like I mentioned in the `GuestMemoryMut` comment, I'm really wondering why that would be necessary, so any examples are greatly appreciated.
- The cost of the `snapshot` operation is significant enough that we want/need to amortize it over multiple inner `GuestMemory` method calls whenever possible.

If either of the above is true, I think this needs to be a part of the main `GuestMemory` interface, because components such as device building blocks that are written without assuming a particular guest memory implementation need to use a uniform interface (I think one of Gerry's comments was also pointing this out). Otherwise, it can be abstracted within a wrapper implementation (such as `GuestMemoryAtomic`) of the regular `GuestMemory` interface.

Comment (bonzini) [reply src/guest_memory.rs]: > The cost of the snapshot operation is significant enough that we want/need to amortize it over multiple inner GuestMemory method calls whenever possible.

It's not really the cost of snapshot(), it's the cost of find_region(). And if you want to cache the result of find_region(), you cannot use the do_in_region() approach and you need a guard. snapshot() provides the guard.

> If either of the above is true, I think this needs to be a part of the main GuestMemory interface

This method is a part of the main interface, though it's a bit roundabout---more precisely,  thanks to the generic implementation of `GuestMemorySnapshottable` it is a part of `&'a GuestMemory` and `Arc<GuestMemory>`, for both of which it can simply return `self.

File: src/guest_memory.rs [hunk 5]
```diff
@@ -235,23 +238,11 @@ pub trait GuestMemoryRegion: Bytes<MemoryRegionAddress, E = Error> {
...
-///
-/// The main responsibilities of the GuestMemory trait are:
-/// - hide the detail of accessing guest's physical address.
-/// - map a request address to a GuestMemoryRegion object and relay the request to it.
-/// - handle cases where an access request spanning two or more GuestMemoryRegion objects.
-///
-/// Note: all regions in a GuestMemory object must not intersect with each other.
-pub trait GuestMemory {
+/// All regions in a GuestMemoryMut object must not intersect with each other.
+pub trait GuestMemoryMut {
```
[the remark this answers is not in the document]
Comment (bonzini) [reply src/guest_memory.rs]: > both traits allow interior mutability regarding the contents of the guest memory regions, but the difference comes from whether the set of regions can be mutated as well (or so I understand).

Yes, this is correct. Mutability of the contents is kind of implicit in the whole `VolatileMemory` concept, what matters here is the set of regions.

> that can be fixed by making find_region return a guard, or switching back to something like the old do_in_region which is based on closures, so nothing has to be leaked to the outside.

You're not very far from the design that I used here, actually. The guard is the `GuestMemory` returned by a `GuestMemoryMut`'s `snapshot()` method; when the `find_region` method is then invoked on the snapshot itself, the lifetime is limited to that of the snapshot (and therefore the snapshot acts as a guard).

> GuestMemoryMut can only guarantee the same over the course of a single method invocation, but why isn't the latter sufficient?

Basically because it prevents caching. In order to implement full-blown caching, something like a `GuestMemoryMut` observer should be added (also mutuated by QEMU's `MemoryListener` concept), but that would be separate because observers in Rust are tricky.

File: src/guest_memory.rs [hunk 6]
```diff
@@ -302,6 +306,96 @@ pub trait GuestMemoryMut {
...
+///     type Memory = &'a GuestMemoryMmap;
+/// }
+/// let mut for_immutable_mmap: VirtioDevice<ImmutableMmap> =
+///     VirtioDevice::new();
+/// let mmap = get_mmap();
+/// for_immutable_mmap.activate(&mmap);
+/// # }
+/// ```
+
+pub trait GuestMemorySnapshottable<M: GuestMemory> {
```
[the remark this answers is not in the document]
Comment (bonzini) [reply src/guest_memory.rs]: What about `GuestAddressSpace`, and renaming the method to `memory_map()`?

File: src/guest_memory.rs (hunk 3 above)
Comment (bonzini) [reply src/guest_memory.rs]: Ok, done.

**Comment by bonzini** [comment]
> Also, I was wondering whether this approach is susceptible to some potential race conditions. For example, if a region is inserted, then an address from the new region is written to an old memory region, and an old snapshot is used to "dereference" the address we'll get an error :(

Yes, this is intended. The behavior of the guest is more or less undefined in that case, so it will be okay as long as the snapshots are used for as little time as possible. Otherwise, you'll get the behavior you mention plus a memory leak. If you do not want this, it's of course possible to implement a `GuestMemoryAtomic` using a Mutex or RwLock.

**Comment by bonzini** [comment]
Regarding naming, in the discussion with Serban I proposed the following replacement:

* GuestMemorySnapshottable -> GuestAddressSpace

* `snapshot()` -> `memory_map()`

File: src/guest_memory.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/guest_memory.rs]: When we are working with Solaris/Linux memory hotplug, we used reference count on pages and observer pattern to reclaim memory on demand. But the whole memory hotplug implementation is really complex, we have struggled with it for years, to fix subsystem/drivers without knowledge about memory hotplug. It's really a hard and long way.

For VMM memory usage patterns, I feel a snapshot mechanism should be a balanced tradeoff between complexity and performance. We only need to maintain region level reference count and address space level snapshots, and changes to GuestMemory consumers to support snapshot should be minor.

For cases which really cares about performance, ArcSwap cache could be used to minimize the   overhead of snapshot, near to zero:)

**Comment by jiangliu** [comment]
> > Also, I was wondering whether this approach is susceptible to some potential race conditions. For example, if a region is inserted, then an address from the new region is written to an old memory region, and an old snapshot is used to "dereference" the address we'll get an error :(
> 
> Yes, this is intended. The behavior of the guest is more or less undefined in that case, so it will be okay as long as the snapshots are used for as little time as possible. Otherwise, you'll get the behavior you mention plus a memory leak. If you do not want this, it's of course possible to implement a `GuestMemoryAtomic` using a Mutex or RwLock.



> > Also, I was wondering whether this approach is susceptible to some potential race conditions. For example, if a region is inserted, then an address from the new region is written to an old memory region, and an old snapshot is used to "dereference" the address we'll get an error :(
> 
> Yes, this is intended. The behavior of the guest is more or less undefined in that case, so it will be okay as long as the snapshots are used for as little time as possible. Otherwise, you'll get the behavior you mention plus a memory leak. If you do not want this, it's of course possible to implement a `GuestMemoryAtomic` using a Mutex or RwLock.

If the memory hotplug controller is implemented correctly, this scenario won't happen. The work flow is something like:
1) Issue memory hotplug command through VMM API
2) VMM prepares new memory resource
3) VMM updates the GuestMemoryMut instance to include the hot-added memory regions with RCU-like mechanism like ArcSwap.
4) VMM waits for a RCU quiesce period to ensure all old snapshots have been released.
5) VMM generates memory hot-add events to guest OS.
6) Guest OS makes use of hot-added memory.
After step 4, all newly taken snapshot should include the hot-added memory.

**Comment by jiangliu** [comment]
> Regarding naming, in the discussion with Serban I proposed the following replacement:
> 
> * GuestMemorySnapshottable -> GuestAddressSpace
> * `snapshot()` -> `memory_map()`

Naming is always hard:)
We have built another abstraction layer over GuestMemory, which already used the notation "AddressSpace", luckily it's not named "GuestAddressSpace".
Actually I have one question here, what's the role of GuestMemory? It's a accessor to access guest memory or it's a resource owner of guest memory. In my point of view, GuestMemory is just an accessor of guest memory, another abstraction of AddressSpace is the resource owner of guest memory. We could generate multiple different GuestMemory objects to access different parts of guest memory. For example, we could split guest memory guest kernel and non-guest-kernel memory, and build one GuestMemory instance for guest kernel and another GuestMemory instance for non-guest-kernel memory.

File: src/guest_memory.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/guest_memory.rs]: > > that can be fixed by making find_region return a guard, or switching back to something like the old do_in_region which is based on closures, so nothing has to be leaked to the outside.
> 
> You're not very far from the design that I used here, actually. The guard is the `GuestMemory` returned by a `GuestMemoryMut`'s `snapshot()` method; when the `find_region` method is then invoked on the snapshot itself, the lifetime is limited to that of the snapshot (and therefore the snapshot acts as a guard).

Yes, the key idea is to return:
1) a reference for non-hotplug-capable guest memory implement to minimize penalty
2) a guard for hotplug-capable guest memory implementation to ensure correctness.
Thanks to rust's lifetime, it's hard to combine a reference and a guard as one type. Paolo and I are trying to solve the lifetime issue with different methods. For C, it's not an issue at all:(

With Paolo's proposal, a possible pattern for device driver may be something like:
```rust
struct DummyDevice<M: GuestMemoryBackend> {
     vm_memory: M::Memory,
}
DummyDevice<M: GuestMemory> {
      pub fn new(vm_memory: M::Memory {
             DummyDevice {
                        vm_memory
             }
      }
      pub fn handler(&self) {
            let snapshot = self.vm_memory.snapshot();
            self.process_request(snapshot);
      }
      fn process_request(&self, snapshot: M:Snapshot) {
      }
}
```
Basically we need to change existing device driver to this mode, but it should be trivial.

[the remark this answers is not in the document]
Comment (jiangliu) [reply src/guest_memory.rs]: Too optimistic, we need to carries on an extra lifetime parameter for GuestMemoryBackend:(

[the remark this answers is not in the document]
Comment (bonzini) [reply src/guest_memory.rs]: Yes, `'a` is still needed but it's much lighter-weight than a generic type parameter, and especially nested generic type parameters which you have to write as `<A, B<A>>`.

**Comment by bonzini** [comment]
> what's the role of GuestMemory? It's a accessor to access guest memory or it's a resource owner of guest memory. In my point of view, GuestMemory is just an accessor of guest memory, another abstraction of AddressSpace is the resource owner of guest memory

Exactly: snapshotting is the mechanism through which the owner grants (time-limited) access to guest memory. When the implementation of GuestMemoryMut and GuestMemory is separate, then you need the GuestMemorySnapshottable trait (or GuestAddressSpace...) to go from one to the other. When it is the same, instead you use the generic implementations on `&'a M`/`Rc<M>`/`Arc<M>`; then access is granted to the client just by passing a reference (which has a well defined lifetime, possibly `'static`, and therefore doubles as a guard) or by cloning a reference-counted pointer (which keeps the resources alive).

**Comment by jiangliu** [comment]
> > what's the role of GuestMemory? It's a accessor to access guest memory or it's a resource owner of guest memory. In my point of view, GuestMemory is just an accessor of guest memory, another abstraction of AddressSpace is the resource owner of guest memory
> 
> Exactly: snapshotting is the mechanism through which the owner grants (time-limited) access to guest memory. When the implementation of GuestMemoryMut and GuestMemory is separate, then you need the GuestMemorySnapshottable trait (or GuestAddressSpace...) to go from one to the other. When it is the same, instead you use the generic implementations on `&'a M`/`Rc<M>`/`Arc<M>`; then access is granted to the client just by passing a reference (which has a well defined lifetime, possibly `'static`, and therefore doubles as a guard) or by cloning a reference-counted pointer (which keeps the resources alive).

Great, we should be on the same page now. My only concern remaining is the extra lifetime parameter. The original design is heavily shaped by the requirement to avoid extra lifetime parameter.

**Comment by bonzini** [comment]
Is it really important? The lifetime parameter is already present if the client simply stores an `& 'a GuestMemoryMmap`, and in the hotplug case it can be replaced by `'_` as in the MutableMmap example from the documentation.

**Comment by jiangliu** [comment]
> Is it really important? The lifetime parameter is already present if the client simply stores an `& 'a GuestMemoryMmap`, and in the hotplug case it can be replaced by `'_` as in the MutableMmap example from the documentation.

I gave myself some more time to think about the overall abstraction:)
When I was starting to work on the vm-memory crate, I built the abstraction as :
AddressSpace as the resource owner and GuestMemory as the accessor, but the AddressSpace concept has been rejected. Then we have moved the AddressSpace into the vmm part.
Now we are talking the concept of AddressSpace again, I really like the concept to split resource owner from accessor:)

I suggest an abstraction as: AddressSpace -> GuestMemoryHandle -> GuestMemorySnapshot.
AddressSpace belongs to vmm, GuestMemoryHandle and GuestMemory belong to vm-memory.
With this abstraction, vmm is the resource owner. Vmm  creates handles and passes them to guest memory clients such as kernel loader/IO devices. Guest memory clients creates stable snapshot view from handles on demand.
Why introduces a layer of GuestMemoryHandle? We want to limit the portion of guest memory accessible to each guest memory clients. For example, we want to constraint the guest memory accessible by IO subsystem.

I have scanned the code with generalized <M: GuestMemory>, the majority usage of generic type parameter `M`  is `&M`, with several instance of Arc<M>. So to ease the code conversion, how about replacing GuestMemoryBackend with GuestMemoryHandle, and keep GuestMemory as is though it's actually GuestMemorySnapshot.

This abstraction should be flexible enough. It could support all of firecracker, cloud hypervisor and our dragonball VMM.

**Comment by jiangliu** [comment]
> Is it really important? The lifetime parameter is already present if the client simply stores an `& 'a GuestMemoryMmap`, and in the hotplug case it can be replaced by `'_` as in the MutableMmap example from the documentation.

For most guest memory clients with generic type <M: GuestMemory>, they store `Arc\<M\>` instead of `&'a M` to avoid dealing with lifetime parameters.

**Comment by bonzini** [comment]
> I suggest an abstraction as: AddressSpace -> GuestMemoryHandle -> GuestMemorySnapshot.

Essentially what you want is something more flexible than GuestMemoryMmap. The last two layers are essentially GuestMemorySnapshottable (now GuestAddressSpace) and GuestMemory. On top of that you want a hierarchy of memory regions similar to QEMU; you can build it in your VMM and wrap it with GuestMemoryAtomic to get the other two for free. It's orthogonal to this issue, and we can even add it to vm-memory under a separate feature.

I tried moving the associated types into GuestMemorySnapshottable but failed, so the backend trait is separate, but I think the system we have is quite usable.

**Comment by jiangliu** [comment]
> > I suggest an abstraction as: AddressSpace -> GuestMemoryHandle -> GuestMemorySnapshot.
> 
> Essentially what you want is something more flexible than GuestMemoryMmap. The last two layers are essentially GuestMemorySnapshottable (now GuestAddressSpace) and GuestMemory. GuestMemoryBackend is a small wrapper around the former.
> 
> On top of that you want a hierarchy of memory regions similar to QEMU; you can build it in your VMM and wrap it with GuestMemoryAtomic to get the other two for free. It's orthogonal to this issue, and we can even add it to vm-memory under a separate feature.
> 
> I tried moving the associated types into GuestMemorySnapshottable but failed, so the backend trait is separate, but I think the system we have is quite usable.

Yes, I like the abstraction. And it could be easily integrated with our existing code.
One minor suggestion is not to call GuestMemoryBackend as GuestAddressSpace, something like GuestMemoryHandle would be welcomed:)

I also tried another solution to add a lifetime parameter to GuestMemory, the implementation is simple, but the abstraction is not as good as your proposal. 
So I hope we could merge this proposal soon:)

File: src/guest_memory.rs [hunk 7]
```diff
@@ -336,6 +340,140 @@ pub trait GuestMemoryMut {
...
+/// # use vm_memory::{GuestMemory, GuestMemoryBackend, GuestMemoryMmap};
+/// struct MmapBackend;
+/// # #[cfg(feature = "backend-mmap")]
+/// impl<'a> GuestMemoryBackend<'a> for MmapBackend {
+///     type Memory = GuestMemoryMmap;
+///     type AddressSpace = &'a GuestMemoryMmap;
+/// }
+/// ```
+
+pub trait GuestMemoryBackend<'a> {
```
Comment (alexandruag) [inline src/guest_memory.rs]: Hmm, why is this trait necessary/helpful in addition to `GuestAddressSpace`?  Is it mostly/mainly a solution to the discussion in #72? I noticed one of Paolo's comments saying some associated type(s) could not be moved in `GuestAddressSpace` so the backend remains separate; what was the desired outcome there?

File: src/guest_memory.rs [hunk 8]
```diff
@@ -336,6 +340,140 @@ pub trait GuestMemoryMut {
...
+///     type AddressSpace = &'a GuestMemoryMmap;
+/// }
+/// let mut for_immutable_mmap: VirtioDevice<ImmutableMmap> =
+///     VirtioDevice::new();
+/// let mmap = get_mmap();
+/// for_immutable_mmap.activate(&mmap);
+/// # }
+/// ```
+
+pub trait GuestAddressSpace<M: GuestMemory> {
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/guest_memory.rs]: I'm working on this way, and it works with a limitation that GuestMemoryMmap will always used as Arc<GuestMemoryMmap>. I think it's a practically acceptable limitation though:)

> Hi again, thanks for the detailed discussions and explanations. Just posted a couple more questions from my side. I'm not super happy with the additional abstraction TBH, but I think we can build most reusable building blocks in a way that depends on `GuestMemory` alone and is agnostic to `GuestAddressSpace`, so it's going to be relatively straightforward to course correct if there's a better /more decoupled way of doing things at a later point.

Actually we could get rid of GuestMemoryMut now.

File: src/guest_memory.rs (hunk 7 above)
Comment (jiangliu) [reply src/guest_memory.rs]: I have worked out a patch upon Paolo's patches to get rid of GuestMemoryBackend/GuestMemoryBackendMT. GuestMemory and GuestAddressSpace should be enough.
By that way we only add an extra layer GuestAddressSpace over GuestMemory.

**Comment by jiangliu** [comment]
>Does GuestMemoryMut still serve any purpose at this point? Seems the model we're heading >towards is a hierarchy where we have things that implement GuestMemory and then the new >AddressSpace concept on top, that provides Deref<Target=M:GuestMemory> guards which we use >to get GuestMemory "leases".

Yes, we could get rid of GuestMemoryMut now with the GuestAddressSpace abstraction.

**Comment by jiangliu** [comment]
Hi @bonzini , I was happening to work on a patch to solve the @alexandruag 's review comments. I have pushed the patch to https://github.com/jiangliu/vm-memory/commit/d5b9b3d648e39e392f489b933142147e48983cc2 , which is based on your patch set.

As @alexandruag has mentioned, we could revert GuestMemoryMut related changes now, eventually we only need to GuestMemroyAddress abstraction to support memory hotplug.

**Comment by bonzini** [comment]
Looks good! I'll try to merge both @jiangliu's refactoring and @alexandruag's other suggested changes (renaming, removing GuestMemoryMut).

**Comment by bonzini** [comment]
> Note this cause a limitation that GuestMemoryMmap will always be used
as Arc<GuestMemoryMmap>, though it's a pratically acceptible limitaiton.

@jiangliu, can you expand on this? This origin of this limitation is not clear to me. It's not a big limitation since you can always `as_ref()` the `Arc`, but I would like to understand it better.

**Comment by jiangliu** [comment]
> > Note this cause a limitation that GuestMemoryMmap will always be used
> > as Arc, though it's a pratically acceptible limitaiton.
> 
> @jiangliu, can you expand on this? This origin of this limitation is not clear to me. It's not a big limitation since you can always `as_ref()` the `Arc`, but I would like to understand it better.

/// # #[cfg(feature = "backend-mmap")]
/// # fn test_1() {
/// // Using `VirtioDevice` with an immutable GuestMemoryMmap:
/// let mut for_immutable_mmap: VirtioDevice<Arc<GuestMemoryMmap>> =
///     VirtioDevice::new();
/// let mmap = Arc::new(get_mmap());
/// for_immutable_mmap.activate(mmap.clone());
/// let mut another: VirtioDevice<Arc<GuestMemoryMmap>> =
///     VirtioDevice::new();
/// another.activate(mmap.clone());
/// # }

The test case fails if we use VirtioDevice<GuestMemoryMmap> instead of VirtioDevice<Arc<GuestMemoryMmap>> because GuestMemoryMmap doesn't implement GuestAddressSpace and it's not easy to implement GuestAddressSpace for GuestMemoryMmap too.
Luckily we have impl<M: GuestMemory> GuestAddressSpace for Arc<M> as Arc<GuestMemoryMmap> works.

**Comment by bonzini** [comment]
Done, this is much nicer. Associated types everywhere, thanks Jiang and Alex for pushing and helping with the simplifications!

**Comment by bonzini** [comment]
@jiangliu Do you have time for a proof of concept of switching vm-virtio to `GuestAddressSpace`? I think we need some kind of real-world testing before committing to the new model.

**Comment by bonzini** [comment]
> The test case fails if we use `VirtioDevice<GuestMemoryMmap>`

Right, you should instead use `VirtioDevice<&GuestMemoryMmap>`. Remember that `GuestAddressSpace` is a replacement for a _reference_ to GuestMemory, not for the GuestMemory itself. So references still work.

**Comment by jiangliu** [comment]
> @jiangliu Do you have time for a proof of concept of switching vm-virtio to `GuestAddressSpace`? I think we need some kind of real-world testing before committing to the new model.

Hi Paolo, I have converted one vm-virtio driver to use the GuestAddressSpace trait, the conversion is strait forward. There are two minor issues left:
1) whether we should mark GuestMemory/GuestAddressSpace as Send/Sync
2) whether we should mark GuestAddressSpace as Clone, we need some experience to choose among Arc<GuestAddressSpace> or GuestAddressSpace: Clone.
But this two minor issues could be handled separately later.

**Comment by jiangliu** [comment]
> > The test case fails if we use `VirtioDevice<GuestMemoryMmap>`
> 
> Right, you should instead use `VirtioDevice<&GuestMemoryMmap>`. Remember that `GuestAddressSpace` is a replacement for a _reference_ to GuestMemory, not for the GuestMemory itself. So references still work.

I tried to make the hotplug/non-hotplug test cases as symmetric as possible, it's an exercise for code reuse:)

> > The test case fails if we use `VirtioDevice<GuestMemoryMmap>`
> 
> Right, you should instead use `VirtioDevice<&GuestMemoryMmap>`. Remember that `GuestAddressSpace` is a replacement for a _reference_ to GuestMemory, not for the GuestMemory itself. So references still work.

I just tried to make the hotplug/non-hotplug test cases as symmetric as possible, it's an exercise for code reuse:)

File: src/atomic.rs [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+impl<M: GuestMemory> GuestMemoryAtomic<M> {
+    /// create a new GuestMemoryAtomic object whose initial contents come from
+    /// the `map` GuestMemory.
+    pub fn new(map: M) -> Self {
+        Arc::new(map).into()
+    }
+
+    fn load(&self) -> Guard<'static, Arc<M>> {
```
Comment (sboeuf) [inline src/atomic.rs]: Aren't we supposed to expose this one as public?

Comment (jiangliu) [reply src/atomic.rs]: Method load() will be hide behind GuestAddressSpace::memory(). External clients do not need direct access to the load method.

Comment (bonzini) [reply src/atomic.rs]: Right, this method only hides the ugly `self.inner.0.`. `memory()` returns the correct wrapper for `arc_swap::Guard`.

Comment (sboeuf) [reply src/atomic.rs]: Yep I realized that after a few min :)

**Comment by sboeuf** [comment]
@jiangliu @bonzini [here](https://github.com/cloud-hypervisor/cloud-hypervisor/pull/762) is the code updated in Cloud-Hypervisor, based on this PR.

**Comment by bonzini** [comment]
@jiangliu @sboeuf can you confirm it's just a testing method? Perhaps making it `pub(crate)` is better.

**Comment by sboeuf** [comment]
@bonzini 
>can you confirm it's just a testing method? Perhaps making it pub(crate) is better.

Which method are you referring to?

**Comment by bonzini** [comment]
Sorry, wrong pull request.

**Comment by sboeuf** [comment]
Just curious, what are we waiting for to get this PR merged now that it seems everybody agreed on the design?

**Comment by bonzini** [comment]
Code review from maintainers I suppose. :)

**Comment by sboeuf** [comment]
@alexandruag @serban300 @andreeaflorescu PTAL :)

**Comment by rbradford** [comment]
@sboeuf needs a rebase before it can be merged
